In [15]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import statistics 
import math

In [16]:
matches = pd.read_csv("/content/sample_data/matches.csv")
matches.head()

id  season       city  ...         umpire1        umpire2 umpire3
0   1    2017  Hyderabad  ...     AY Dandekar       NJ Llong     NaN
1   2    2017       Pune  ...  A Nand Kishore         S Ravi     NaN
2   3    2017     Rajkot  ...     Nitin Menon      CK Nandan     NaN
3   4    2017     Indore  ...    AK Chaudhary  C Shamshuddin     NaN
4   5    2017  Bangalore  ...             NaN            NaN     NaN

[5 rows x 18 columns]

In [17]:
# setting up the table with relevant columns
dropList = ["result","player_of_match","venue","umpire1","umpire2","umpire3"]
matches.drop(labels=dropList, axis=1, inplace=True)
matches.head() 

id  season  ... win_by_runs win_by_wickets
0   1    2017  ...          35              0
1   2    2017  ...           0              7
2   3    2017  ...           0             10
3   4    2017  ...           0              6
4   5    2017  ...          15              0

[5 rows x 12 columns]

In [18]:
matches[pd.isnull(matches['winner'])]

id  season       city  ... winner win_by_runs win_by_wickets
300    301    2011      Delhi  ...    NaN           0              0
545    546    2015  Bangalore  ...    NaN           0              0
570    571    2015  Bangalore  ...    NaN           0              0
744  11340    2019  Bengaluru  ...    NaN           0              0

[4 rows x 12 columns]

In [19]:
matches['winner'].fillna('Draw', inplace=True)
matches[pd.isnull(matches['winner'])]

Empty DataFrame
Columns: [id, season, city, date, team1, team2, toss_winner, toss_decision, dl_applied, winner, win_by_runs, win_by_wickets]
Index: []

In [20]:
matches[pd.isnull(matches['city'])]

id  season city  ...                       winner win_by_runs win_by_wickets
461  462    2014  NaN  ...  Royal Challengers Bangalore           0              7
462  463    2014  NaN  ...             Delhi Daredevils           0              4
466  467    2014  NaN  ...          Chennai Super Kings           7              0
468  469    2014  NaN  ...          Sunrisers Hyderabad           4              0
469  470    2014  NaN  ...          Chennai Super Kings           0              7
474  475    2014  NaN  ...              Kings XI Punjab           0              5
476  477    2014  NaN  ...          Sunrisers Hyderabad          15              0

[7 rows x 12 columns]

In [21]:
matches['city'].fillna('Dubai', inplace=True)
matches[pd.isnull(matches['city'])]

Empty DataFrame
Columns: [id, season, city, date, team1, team2, toss_winner, toss_decision, dl_applied, winner, win_by_runs, win_by_wickets]
Index: []

In [22]:
matches.replace(["Deccan Chargers","Delhi Daredevils"],["Sunrisers Hyderabad","Delhi Capitals"],inplace=True,regex=True)
match1 = matches[((matches.team1=="Sunrisers Hyderabad") & (matches.team2=="Kings XI Punjab")) | ((matches.team1=="Kings XI Punjab") & (matches.team2=="Sunrisers Hyderabad"))]
match1.shape[0]

24

In [23]:
mw_kxip = 0 
mw_srh = 0
lst= [i for i in match1['winner']] 
print("Win Tracker!")
for i in lst:
  if i=='Sunrisers Hyderabad':
    mw_srh += 1
  else:
    mw_kxip += 1
  print(str(mw_srh)+" "+str(mw_kxip))
print("SRH vs KXIP : "+str(mw_srh)+" "+str(mw_kxip))

Win Tracker!
1 0
2 0
2 1
2 2
2 3
2 4
3 4
4 4
4 5
5 5
5 6
5 7
6 7
7 7
7 8
7 9
8 9
9 9
10 9
11 9
11 10
12 10
12 11
13 11
SRH vs KXIP : 13 11


In [24]:
last_3_season = matches[(matches['season']>=2017) & np.logical_or(np.logical_and(matches.team1=="Sunrisers Hyderabad",matches.team2=="Kings XI Punjab"),np.logical_and(matches.team1=="Kings XI Punjab",matches.team2=="Sunrisers Hyderabad"))]
last_3_season.groupby('winner').season.count()

winner
Kings XI Punjab        2
Sunrisers Hyderabad    4
Name: season, dtype: int64

Out of 28 matches held between KXIP and SRH, SRH has edge majority of 13 wins overall.Same is the situation in past three seasons, SRH has steady record of 4 by 2 over KXIP.

In [25]:
deliveries = pd.read_csv("/content/sample_data/deliveries.csv")
deliveries.head()

match_id  inning  ... dismissal_kind fielder
0         1       1  ...            NaN     NaN
1         1       1  ...            NaN     NaN
2         1       1  ...            NaN     NaN
3         1       1  ...            NaN     NaN
4         1       1  ...            NaN     NaN

[5 rows x 21 columns]

In [26]:
dropToBeList = ['inning','is_super_over','bye_runs','legbye_runs','fielder']
deliveries.drop(dropToBeList, axis=1, inplace=True)
deliveries['dismissal_kind'].fillna('Not Out',inplace=True)
deliveries.head()

match_id         batting_team  ... player_dismissed  dismissal_kind
0         1  Sunrisers Hyderabad  ...              NaN         Not Out
1         1  Sunrisers Hyderabad  ...              NaN         Not Out
2         1  Sunrisers Hyderabad  ...              NaN         Not Out
3         1  Sunrisers Hyderabad  ...              NaN         Not Out
4         1  Sunrisers Hyderabad  ...              NaN         Not Out

[5 rows x 16 columns]

In [27]:
ballbyball = deliveries[((deliveries.batting_team=="Sunrisers Hyderabad") & (deliveries.bowling_team=="Kings XI Punjab")) | ((deliveries.batting_team=="Kings XI Punjab") & (deliveries.bowling_team=="Sunrisers Hyderabad"))]
ballbyball.head()

match_id         batting_team  ... player_dismissed  dismissal_kind
4312        19  Sunrisers Hyderabad  ...              NaN         Not Out
4313        19  Sunrisers Hyderabad  ...              NaN         Not Out
4314        19  Sunrisers Hyderabad  ...              NaN         Not Out
4315        19  Sunrisers Hyderabad  ...              NaN         Not Out
4316        19  Sunrisers Hyderabad  ...              NaN         Not Out

[5 rows x 16 columns]

In [28]:
#Q5
ub = ballbyball[ballbyball['batsman']=='CH Gayle']
matches_played = [i for i in ub['match_id']]
matches_played = list(set(matches_played))
the_lst = []
for i in matches_played:
  df = ub[ub['match_id']==i]
  lst = [i for i in df['batsman_runs']]
  if(len(lst)<=10):
    score = sum(lst)
    the_lst.append(score)
  else:
    lst = lst[:10]
    score = sum(lst)
    the_lst.append(score)
print(len(matches_played))
print(the_lst)
print(statistics.median(the_lst))
# vs SRH

4
[15, 4, 17, 10]
12.5


Against SRH in KXIP, he played 4 matches and smashed 12.5 runs in the first ten deliveries he face!

In [29]:
#Q5
ub = deliveries[(deliveries['batsman']=='CH Gayle') & (deliveries['bowling_team']=="Sunrisers Hyderabad")]
matches_played = [i for i in ub['match_id']]
matches_played = list(set(matches_played))
the_lst = []
for i in matches_played:
  df = ub[ub['match_id']==i]
  lst = [i for i in df['batsman_runs']]
  if(len(lst)<=10):
    score = sum(lst)
    the_lst.append(score)
  else:
    lst = lst[:10]
    score = sum(lst)
    the_lst.append(score)
print(len(matches_played))
print(the_lst)
print(statistics.median(the_lst))
# in IPL

13
[21, 9, 11, 1, 17, 4, 13, 10, 6, 15, 6, 35, 11]
11


In IPL, in first ten balls he averagely scored 11 runs which is far better than his ipl average calculated(8.5)

In [30]:
#Q2
dw= deliveries[deliveries['batsman']=="DA Warner"]
dw_dismissal_stats = dw.groupby('dismissal_kind').dismissal_kind.count()
dw_matches_played = list(set([i for i in dw['match_id']]))
print(len(dw_matches_played))
print(dw_dismissal_stats)
runs_scored = dw.groupby('match_id').batsman_runs.sum()
balls_faced = dw.groupby('match_id').ball.count()
strike_rate = (runs_scored/balls_faced)*100
strike_rate.describe() # in IPL

126
dismissal_kind
Not Out              3284
bowled                 23
caught                 69
caught and bowled       2
hit wicket              1
lbw                     4
run out                10
stumped                 5
Name: dismissal_kind, dtype: int64


count    126.000000
mean     120.454240
std       53.993947
min        0.000000
25%       89.523810
50%      124.404762
75%      157.506053
max      223.076923
dtype: float64

Out of 126 matches played by Warner in his IPL career, he has been dismissed 114 times (i.e) 91% of times. and He maintains a strike rate between 120-125 on an average!

In [31]:
#Q2
dw= deliveries[(deliveries['batsman']=="DA Warner") & (deliveries['bowling_team']=="Kings XI Punjab")] 
dw_dismissal_stats = dw.groupby('dismissal_kind').dismissal_kind.count()
dw_matches_played = list(set([i for i in dw['match_id']]))
print(len(dw_matches_played))
print(dw_dismissal_stats)
runs_scored = dw.groupby('match_id').batsman_runs.sum()
balls_faced = dw.groupby('match_id').ball.count()
strike_rate = (runs_scored/balls_faced)*100
strike_rate.describe() # vs KXIP

17
dismissal_kind
Not Out       578
bowled          2
caught         10
hit wicket      1
run out         4
Name: dismissal_kind, dtype: int64


count     17.000000
mean     127.937397
std       50.729548
min        0.000000
25%      111.111111
50%      141.463415
75%      155.555556
max      188.888889
dtype: float64

on KXIP, he has been dismissed 100% of times and has a good strike between 130-140 !

In [32]:
#Q3
kl_rahul = deliveries[deliveries['batsman']=="KL Rahul"]
total_runs_scored = kl_rahul.groupby('match_id').batsman_runs.sum()
kl_rahul_scores = [i for i in total_runs_scored]
#print(kl_rahul_scores)
teams_he_played_for = pd.DataFrame(kl_rahul.groupby('match_id').batting_team.unique())
#print(teams_he_played_for)
matches_he_played = []
for i in kl_rahul['match_id']:
  if i not in matches_he_played:
    matches_he_played.append(i)
teams_he_played = [i[0] for i in teams_he_played_for['batting_team']]
team_scores=[]
for i in range(len(matches_he_played)):
  total_score_by_team = deliveries[(deliveries.batting_team==teams_he_played[i]) & (deliveries.match_id==matches_he_played[i])]
  score_by_ball = [i for i in total_score_by_team['total_runs']]
  team_scores.append(sum(score_by_ball))
perc=[]
for i in range(len(kl_rahul_scores)):
  val = kl_rahul_scores[i]/team_scores[i]
  perc.append(round(val*100))
print(statistics.median(perc))
print(statistics.mean(perc))

18.5
21.051724137931036


In [38]:
#Q3
kl_rahul = ballbyball[(ballbyball['batsman']=="KL Rahul") & (ballbyball['batting_team']=="Kings XI Punjab")]
total_runs_scored = kl_rahul.groupby('match_id').batsman_runs.sum()
kl_rahul_scores = [i for i in total_runs_scored]
#print(kl_rahul_scores)
teams_he_played_for = pd.DataFrame(kl_rahul.groupby('match_id').batting_team.unique())
#print(teams_he_played_for)
matches_he_played = []
for i in kl_rahul['match_id']:
  if i not in matches_he_played:
    matches_he_played.append(i)
teams_he_played = [i[0] for i in teams_he_played_for['batting_team']]
team_scores=[]
for i in range(len(matches_he_played)):
  total_score_by_team = deliveries[(deliveries.batting_team==teams_he_played[i]) & (deliveries.match_id==matches_he_played[i])]
  score_by_ball = [i for i in total_score_by_team['total_runs']]
  team_scores.append(sum(score_by_ball))
perc=[]
for i in range(len(kl_rahul_scores)):
  val = kl_rahul_scores[i]/team_scores[i]
  perc.append(round(val*100))
print(statistics.median(perc))
print(statistics.mean(perc))

36.5
32.5


On an average, Rahul contributes to 18-21% of runs to the team score!

In [34]:
#Q4
srh = deliveries[deliveries['batting_team']=="Sunrisers Hyderabad"]
srh_match_ids=[]
for i in srh['match_id']:
  if i not in srh_match_ids:
    srh_match_ids.append(i)
balls_to_fifty=[]
for i in srh_match_ids:
  df = srh[srh['match_id']==i]
  ball_by_ball=[i for i in df['total_runs']]
  score=0
  ball=0
  for j in ball_by_ball:
    score += j
    if(score<=50):
      ball += 1
      continue
    else:
      score -= j
      balls_to_fifty.append(ball)
      break
print(statistics.median(balls_to_fifty))
print(statistics.mean(balls_to_fifty))
# in IPL

41
41.177570093457945


SRH usually take 3 or 4 balls more than matches with KXIP to reach their 50 mark!

In [35]:
#Q4
srh = ballbyball[ballbyball['batting_team']=="Sunrisers Hyderabad"]
srh_match_ids=[]
for i in srh['match_id']:
  if i not in srh_match_ids:
    srh_match_ids.append(i)
balls_to_fifty=[]
for i in srh_match_ids:
  df = srh[srh['match_id']==i]
  ball_by_ball=[i for i in df['total_runs']]
  score=0
  ball=0
  for j in ball_by_ball:
    score += j
    if(score<=50):
      ball += 1
      continue
    else:
      score -= j
      balls_to_fifty.append(ball)
      break
print(statistics.median(balls_to_fifty))
print(statistics.mean(balls_to_fifty))
# vs KXIP

38.5
39.642857142857146
